infra-in-containers/collab.md

```diff
@@ -0,0 +1,430 @@
+
+The customised image should be built. Try to run this image and test if the container is working with your configuration.
+
+### Push image to personal Docker Hub
+
+The customised image can be push to a personal Docker Hub account. Thereafter, this will be available remotely for pulling. To do this, login into your docker hub account on your machine. The command to login is `docker login`. Enter your username and password when prompt.
+
+You may also commit a container to an image which you ran previously. use `docker ps -a` to check the container ID of the container you want to commit. Next use the command `docker commit containerID imagename`. The new image will be created and can be verified by using `docker images`. You can also do `docker tag` to give a more detailed tag on the image.
+Lastly, use `docker push imagename` and the image will be push to your personal Docker Hub.
+
+# Spin off an Email Server with Containers
+## Using Docker Hub Images and Docker Compose
+A readily deployable email server suite is used.
+The relevant docker images from public repository are:
+1. tvial/docker-mailserver
+1. hardware/rainloop
+1. abiosoft/caddy
+
+Docker-mailserver consists of Postfix, Dovecot, Spam Assassin etc. to form the email service, Rainloop serves as the web client for the email service, and Caddy is the web server.
+
+Make a directory for your docker project
+```bash
+mkdir ~/dockerproj
+```
+
+Also make a directory for the email server suite under the docker project directory
+```bash
+mkdir ~/dockerproj/docker-mailserver
+```
+
+Pull the Container Images from Docker Hub into the email server suite directory
+```bash
+cd ~/dockerproj/docker-mailserver
+docker pull tvial/docker-mailserver
+docker pull hardware/rainloop
+docker pull abiosoft/caddy
+```
+
+Also download tvial/docker-mailserver's configuration files
+```bash
+curl -o setup.sh https://raw.githubusercontent.com/tomav/docker-mailserver/master/setup.sh;
+chmod a+x ./setup.sh;
+curl -o docker-compose.yml https://raw.githubusercontent.com/tomav/docker-mailserver/master/docker-compose.yml.dist
+curl -o .env https://raw.githubusercontent.com/tomav/docker-mailserver/master/.env.dist
+```
+
+Edit some of the environment parameters
+```bash
+vim ~/dockerproj/docker-mailserver/.env
+
+  HOSTNAME=mail
+  DOMAINNAME=commandocloudlet.com
+  CONTAINER_NAME=mail
+  SSL_TYPE=manual
+```
+
+Edit the email server suite's docker compose startup file
+```bash
+vim ~/dockerproj/docker-mailserver/docker-compose.yml
+
+  version: '3'
+  services:
+    rainloop:
+      image: hardware/rainloop
+      links:
+      - mail
+      volumes:
+      - ./data/rainloop:/rainloop/data
+    mail:
+      image: tvial/docker-mailserver:latest
+      hostname: ${HOSTNAME}
+      domainname: ${DOMAINNAME}
+      container_name: ${CONTAINER_NAME}
+      ports:
+      - "25:25"
+      - "143:143"
+      - "465:465"
+      - "587:587"
+      - "993:993"
+      - "4190:4190"
+      volumes:
```

Comment:
While this approach isn't wrong, using a driver is better for data persistence. This example uses Docker's local store.
```
volumes:
  postfix_data:
    driver: local
```

The troublesome part is that you have to enter the container to access the files. However, this is a better approach when you have network storage (eg. NFS, glusterfs, ceph).